neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/RemoteNeo4jDialect.java

```diff
@@ -0,0 +1,559 @@
+import org.hibernate.ogm.datastore.neo4j.remote.dialect.impl.RemoteNeo4jTupleAssociationSnapshot;
+import org.hibernate.ogm.datastore.neo4j.remote.dialect.impl.RemoteNeo4jTupleSnapshot;
+import org.hibernate.ogm.datastore.neo4j.remote.dialect.impl.RemoteNodesTupleIterator;
+import org.hibernate.ogm.datastore.neo4j.remote.dialect.impl.RemoteNeo4jSequenceGenerator;
+import org.hibernate.ogm.datastore.neo4j.remote.impl.Neo4jClient;
+import org.hibernate.ogm.datastore.neo4j.remote.impl.RemoteNeo4jDatastoreProvider;
+import org.hibernate.ogm.datastore.neo4j.remote.json.impl.ErrorResponse;
+import org.hibernate.ogm.datastore.neo4j.remote.json.impl.Graph.Node;
+import org.hibernate.ogm.datastore.neo4j.remote.json.impl.Graph.Relationship;
+import org.hibernate.ogm.datastore.neo4j.remote.json.impl.Statement;
+import org.hibernate.ogm.datastore.neo4j.remote.json.impl.Statements;
+import org.hibernate.ogm.datastore.neo4j.remote.json.impl.StatementsResponse;
+import org.hibernate.ogm.dialect.query.spi.BackendQuery;
+import org.hibernate.ogm.dialect.query.spi.ClosableIterator;
+import org.hibernate.ogm.dialect.query.spi.QueryParameters;
+import org.hibernate.ogm.dialect.spi.AssociationContext;
+import org.hibernate.ogm.dialect.spi.ModelConsumer;
+import org.hibernate.ogm.dialect.spi.NextValueRequest;
+import org.hibernate.ogm.dialect.spi.TransactionContext;
+import org.hibernate.ogm.dialect.spi.TupleAlreadyExistsException;
+import org.hibernate.ogm.dialect.spi.TupleContext;
+import org.hibernate.ogm.model.key.spi.AssociatedEntityKeyMetadata;
+import org.hibernate.ogm.model.key.spi.AssociationKey;
+import org.hibernate.ogm.model.key.spi.AssociationKeyMetadata;
+import org.hibernate.ogm.model.key.spi.EntityKey;
+import org.hibernate.ogm.model.key.spi.EntityKeyMetadata;
+import org.hibernate.ogm.model.key.spi.RowKey;
+import org.hibernate.ogm.model.spi.Association;
+import org.hibernate.ogm.model.spi.AssociationOperation;
+import org.hibernate.ogm.model.spi.Tuple;
+import org.hibernate.ogm.model.spi.TupleOperation;
+import org.hibernate.ogm.persister.impl.OgmCollectionPersister;
+import org.hibernate.ogm.persister.impl.OgmEntityPersister;
+import org.hibernate.persister.collection.CollectionPersister;
+import org.hibernate.persister.entity.EntityPersister;
+
+/**
+ * Abstracts Hibernate OGM from Neo4j.
+ * <p>
+ * A {@link Tuple} is saved as a {@link Node} where the columns are converted into properties of the node.<br>
+ * An {@link Association} is converted into a {@link Relationship} identified by the {@link AssociationKey} and the
+ * {@link RowKey}. The type of the relationship is the value returned by
+ * {@link AssociationKeyMetadata#getCollectionRole()}.
+ * <p>
+ * If the value of a property is set to null the property will be removed (Neo4j does not allow to store null values).
+ *
+ * @author Davide D'Alto &lt;davide@hibernate.org&gt;
+ */
+public class RemoteNeo4jDialect extends BaseNeo4jDialect {
```

Comment:
I'm still a bit worried about this class. It duplicates quite a lot of things.

Have you tried to use generics on BaseNeo4jDialect with a couple of methods like newEntityQueries(),... , newTupleSnapshot() and so on?

Not sure it's a good idea but it might be worth a try if you haven't already done so.